### PR TITLE
Feature/new firmware refactor passing all the stuff around until crazy

### PR DESCRIPTION
**Pull request by agutenkunst**
Removes all `shared_ptr` and `unique_ptr`. 
Starts using `FRIEND_TEST`

TODO:
- [x] Cleanup
- [x] Review (especially tests)

### CODE REVIEW

File: include/psen_scan/scanner.h [hunk 1]
```diff
@@ -35,33 +35,55 @@ class LaserScanBuildFailure : public std::runtime_error
...
   void start();
   void stop();
   LaserScan getCompleteScan();
 
 private:
-  std::shared_ptr<ScannerController> scanner_controller_;
+  SC scanner_controller_;
+
+  friend class ScannerTest;
+  FRIEND_TEST(ScannerTest, testConstructorSuccess);
```
Comment (martiniil) [inline include/psen_scan/scanner.h:53]: `#include <gtest/gtest_prod.h>` missing

File: include/psen_scan/ros_scanner_node.h [hunk 2]
```diff
@@ -30,26 +34,127 @@ namespace psen_scan
  * @brief Class implements a ROS-Node for the PSENscan safety laser scanner
  *
  */
-class ROSScannerNode
+template <typename S = Scanner>
+class ROSScannerNodeImplT
```
Comment (martiniil) [inline include/psen_scan/ros_scanner_node.h]: Why not name it `ROSScannerNodeT`?

File: src/psen_scan_driver.cpp [hunk 3]
```diff
@@ -44,40 +44,11 @@ int main(int argc, char** argv)
...
-    // binding callbacks
-    send_start_request_callback = std::bind(&ScannerControllerImpl::sendStartRequest, &scanner_controller);
-    error_callback = std::bind(&ScannerControllerImpl::handleError, &scanner_controller, std::placeholders::_1);
-
-    std::unique_ptr<ScannerImpl> scanner{ new ScannerImpl(
-        std::make_shared<ScannerControllerImpl>(scanner_controller)) };
-
-    ROSScannerNode ros_scanner_node(
-        pnh, DEFAULT_PUBLISH_TOPIC, param_handler.getFrameID(), param_handler.getXAxisRotation(), std::move(scanner));
+    ROSScannerNode ros_scanner_node(pnh,
```
Comment (martiniil) [inline src/psen_scan_driver.cpp:49]: missing header: ros.h

Comment (martiniil) [inline src/psen_scan_driver.cpp:49]: :+1: for clean code

File: test/unit_tests/unittest_scanner.cpp [hunk 4]
```diff
@@ -20,50 +20,58 @@
...
 
-namespace psen_scan_test
+namespace psen_scan
 {
+static const std::string HOST_IP{ "127.0.0.1" };
+static constexpr int HOST_UDP_PORT_DATA{ 50505 };
+static constexpr int HOST_UDP_PORT_CONTROL{ 55055 };
+static const std::string DEVICE_IP{ "127.0.0.100" };
+static constexpr double START_ANGLE{ 0. };
+static constexpr double END_ANGLE{ 275. * 2 * M_PI / 360. };
```
Comment (martiniil) [inline test/unit_tests/unittest_scanner.cpp]: You could use `degreeToRad()` here.

**Review by martiniil — CHANGES_REQUESTED** [review]
Some minor comments. In general, I approve the new concept.

**Comment by ct2034** [comment]
tested on hardware ✅

File: include/psen_scan/ros_scanner_node.h (hunk 2 above)
Comment (ct2034) [reply include/psen_scan/ros_scanner_node.h]: According to https://www.bfilipek.com/2018/01/pimpl.html `ROSScannerNodeImpl` would be sufficient

File: include/psen_scan/scanner.h [hunk 5]
```diff
@@ -35,33 +35,55 @@ class LaserScanBuildFailure : public std::runtime_error
   LaserScanBuildFailure(const std::string& msg = "Error while building laser scan");
 };
 
-// LCOV_EXCL_START
-class Scanner
+template <typename SC = ScannerController>
+class ScannerT
```
Comment (ct2034) [inline include/psen_scan/scanner.h:41]: Please remove the T.

File: include/psen_scan/scanner_controller.h [hunk 6]
```diff
@@ -40,25 +41,11 @@ static constexpr std::chrono::milliseconds RECEIVE_TIMEOUT{ 1000 };
 
 static constexpr uint32_t DEFAULT_SEQ_NUMBER{ 0 };
 
-// LCOV_EXCL_START
-class ScannerController
+template <typename SM = ControllerStateMachine, typename UDPC = UdpClientImpl>
```
Comment (ct2034) [inline include/psen_scan/scanner_controller.h]:
```suggestion
template <typename TCSM = ControllerStateMachine, typename TUCI = UdpClientImpl>
```

File: include/psen_scan/start_request.h [hunk 7]
```diff
@@ -38,8 +38,8 @@ class StartRequest
 
   uint32_t getCRC() const;
 
-  using RawType = std::array<char, START_REQUEST_SIZE>;
-  RawType toCharArray() const;
+  using RawType = std::vector<char>;
+  RawType toRaw() const;
```
Comment (ct2034) [inline include/psen_scan/start_request.h]:
```suggestion
  RawType toRawType() const;
```

File: include/psen_scan/udp_client.h [hunk 8]
```diff
@@ -40,17 +40,17 @@ namespace psen_scan
...
 using ErrorHandler = std::function<void(const std::string&)>;
 
-class UdpClient
-{
-public:
-  virtual ~UdpClient() = default;
+// class UdpClient
+// {
+// public:
+//   virtual ~UdpClient() = default;
```
Comment (ct2034) [inline include/psen_scan/udp_client.h]:
```suggestion
```

File: include/psen_scan/udp_client.h [hunk 9]
```diff
@@ -40,17 +40,17 @@ namespace psen_scan
...
 
-  virtual void startReceiving(const std::chrono::high_resolution_clock::duration timeout) = 0;
-  virtual void write(std::shared_ptr<char> data, const std::size_t& number_of_bytes) = 0;
-  virtual void close() = 0;
-};
+//   virtual void startReceiving(const std::chrono::high_resolution_clock::duration timeout) = 0;
+//   virtual void write(std::shared_ptr<char> data, const std::size_t& number_of_bytes) = 0;
+//   virtual void close() = 0;
+// };
```
Comment (ct2034) [inline include/psen_scan/udp_client.h]:
```suggestion
```

File: src/start_request.cpp [hunk 10]
```diff
@@ -43,14 +43,14 @@ uint32_t StartRequest::getCRC() const
 {
   boost::crc_32_type result;
 
-  std::array<char, START_REQUEST_SIZE> char_array = toCharArray();
+  std::vector<char> char_array = toRaw();
```
Comment (ct2034) [inline src/start_request.cpp]:
```suggestion
  std::vector<char> char_vector = toRaw();
```

File: src/start_request.cpp [hunk 11]
```diff
@@ -89,14 +89,21 @@ StartRequest::RawType StartRequest::toCharArray() const
     os.write((char*)&slave_resolution, sizeof(slave_resolution));
   }
 
-  StartRequest::RawType ret_val{};
+  // StartRequest::RawType ret_val{};
```
Comment (ct2034) [inline src/start_request.cpp]:
```suggestion
```

File: src/start_request.cpp [hunk 12]
```diff
@@ -89,14 +89,21 @@ StartRequest::RawType StartRequest::toCharArray() const
...
 
   // TODO check limits
   std::string data_str(os.str());
-  assert(data_str.length() == START_REQUEST_SIZE);
-  std::copy(data_str.begin(), data_str.end(), ret_val.begin());
+  // assert(data_str.length() == START_REQUEST_SIZE);
+  // std::copy(data_str.begin(), data_str.end(), ret_val.begin());
```
Comment (ct2034) [inline src/start_request.cpp]:
```suggestion
```

File: src/start_request.cpp [hunk 13]
```diff
@@ -89,14 +89,21 @@ StartRequest::RawType StartRequest::toCharArray() const
...
 
-  return ret_val;
+  std::vector<char> v;
+  v.reserve(data_str.length());
+
+  std::copy(data_str.begin(), data_str.end(), std::back_inserter(v));
+  // for(size_t i = 0; i < data_str.length(); ++i)
+  //   std::cout << (int) v[i] << "    " << (int) ret_val[i] << "\n";
```
Comment (ct2034) [inline src/start_request.cpp]:
```suggestion
```

File: src/start_request.cpp [hunk 14]
```diff
@@ -89,14 +89,21 @@ StartRequest::RawType StartRequest::toCharArray() const
     os.write((char*)&slave_resolution, sizeof(slave_resolution));
   }
 
-  StartRequest::RawType ret_val{};
+  // StartRequest::RawType ret_val{};
 
   // TODO check limits
   std::string data_str(os.str());
-  assert(data_str.length() == START_REQUEST_SIZE);
```
Comment (ct2034) [inline src/start_request.cpp:96]: should we add the assertion again?

File: test/unit_tests/unittest_start_request.cpp [hunk 15]
```diff
@@ -35,7 +35,7 @@ ::testing::AssertionResult
 DecodingEquals(StartRequest::RawType const& data, std::size_t offset, T expected, const Endian& endian = Endian::LITTLE)
 {
   T actual_val;
-  memcpy(&actual_val, (char*)&data + offset, sizeof(T));
+  memcpy(&actual_val, &data.front() + offset, sizeof(T));
```
Comment (ct2034) [inline test/unit_tests/unittest_start_request.cpp]:
```suggestion
  memcpy(&actual_val, &data.data() + offset, sizeof(T));
```
https://stackoverflow.com/a/10740266/1493204

File: include/psen_scan/udp_client.h [hunk 16]
```diff
@@ -179,13 +179,13 @@ inline void UdpClientImpl::sendCompleteHandler(const boost::system::error_code&
...
+  std::cout << "Data successfully sent." << std::endl;
 }
 
-inline void UdpClientImpl::write(std::shared_ptr<char> data, const std::size_t& number_of_bytes)
+inline void UdpClientImpl::write(const std::vector<char>& data)
 {
-  io_service_.post([this, data, number_of_bytes]() {
-    socket_.async_send(boost::asio::buffer(data.get(), number_of_bytes),
+  io_service_.post([this, data]() {
+    socket_.async_send(boost::asio::buffer(&data.front(), data.size()),
```
Comment (ct2034) [inline include/psen_scan/udp_client.h]:
```suggestion
    socket_.async_send(boost::asio::buffer(&data.data(), data.size()),
```
https://stackoverflow.com/a/10740266/1493204

**Review by ct2034 — CHANGES_REQUESTED** [review]

File: include/psen_scan/scanner.h (hunk 5 above)
Comment (agutenkunst) [reply include/psen_scan/scanner.h:41]: Will not work

File: test/unit_tests/unittest_start_request.cpp (hunk 15 above)
Comment (agutenkunst) [reply test/unit_tests/unittest_start_request.cpp]: Done

File: include/psen_scan/udp_client.h (hunk 16 above)
Comment (agutenkunst) [reply include/psen_scan/udp_client.h]: Done.

File: src/start_request.cpp (hunk 13 above)
Comment (agutenkunst) [reply src/start_request.cpp]: Done

**Review by ct2034 — APPROVED** [review]
approved if my comments are adressed

**Pull request merged**